### PR TITLE
KAFKA-17342: Moved common coordinator code to separate module.

### PR DESCRIPTION
**Pull request by smjn**
There is a lot of code in `group-coordinator` which is not share/consumer/classic group specific. 

Since we are introducing a `share-coordinator` as part of KIP-932 (in a new module), it would make sense to get the common coordinator functionality into a separate common coordinator module so that `share-coordinator` need not depend on `group-coordinator`.

### CODE REVIEW

File: coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntimeTest.java [hunk 1]
```diff
@@ -629,8 +629,8 @@ public void testScheduleLoading() {
                 .withEventProcessor(new DirectEventProcessor())
                 .withPartitionWriter(writer)
                 .withCoordinatorShardBuilderSupplier(supplier)
-                .withCoordinatorRuntimeMetrics(mock(GroupCoordinatorRuntimeMetrics.class))
-                .withCoordinatorMetrics(mock(GroupCoordinatorMetrics.class))
+                .withCoordinatorRuntimeMetrics(Mockito.mock(CoordinatorRuntimeMetrics.class))
+                .withCoordinatorMetrics(Mockito.mock(CoordinatorMetrics.class))
```
Comment (jeffkbkim) [inline coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntimeTest.java]: for all of the mock related changes, can't we use `import static org.mockito.Mockito.when;` in L103?

**Review by apoorvmittal10 — APPROVED** [review]
Thanks for the PR. LGTM!

File: coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/MultiThreadedEventProcessorTest.java [hunk 2]
```diff
@@ -45,7 +46,6 @@
 import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.Mockito.doAnswer;
-import static org.mockito.Mockito.mock;
```
Comment (jeffkbkim) [inline coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/MultiThreadedEventProcessorTest.java]: any reason we're changing this?

File: coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/SnapshottableCoordinatorTest.java [hunk 3]
```diff
@@ -14,17 +14,16 @@
...
 import org.apache.kafka.common.TopicPartition;
 import org.apache.kafka.common.utils.LogContext;
 import org.apache.kafka.common.utils.MockTime;
-import org.apache.kafka.coordinator.group.MockCoordinatorTimer;
+import org.apache.kafka.coordinator.common.MockCoordinatorTimer;
 import org.apache.kafka.timeline.SnapshotRegistry;
 
 import org.junit.jupiter.api.Test;
 
-import static org.apache.kafka.coordinator.group.runtime.CoordinatorRuntimeTest.MockCoordinatorShard;
```
Comment (jeffkbkim) [inline coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/SnapshottableCoordinatorTest.java:27]: ditto here

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java [hunk 4]
```diff
@@ -3117,11 +3117,11 @@ public void testSessionTimeoutExpiration() {
         context.assertSessionTimeout(groupId, memberId, 45000);
 
         // Advance time past the session timeout.
-        List<ExpiredTimeout<Void, CoordinatorRecord>> timeouts = context.sleep(45000 + 1);
+        List<MockCoordinatorTimer.ExpiredTimeout<Void, CoordinatorRecord>> timeouts = context.sleep(45000 + 1);
```
Comment (jeffkbkim) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java]: can we import
```
import org.apache.kafka.coordinator.group.MockCoordinatorTimer.ExpiredTimeout;
import org.apache.kafka.coordinator.group.MockCoordinatorTimer.ScheduledTimeout;
```
instead? some of the lines become too long

File: settings.gradle [hunk 5]
```diff
@@ -57,6 +57,7 @@ include 'clients',
     'connect:runtime',
     'connect:test-plugins',
     'connect:transforms',
+    'coordinator-common',
```
Comment (jeffkbkim) [inline settings.gradle:60]: can you help me understand what this does? my understanding is we should only disclose group-coordinator-api to the clients module

**Review by jeffkbkim — COMMENTED** [review]
thanks! left some comments.

also, i was wondering if CoordinatorRecordSerde and CoordinatorRecordSerde should now be prefixed with `Group`

File: coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/MultiThreadedEventProcessorTest.java (hunk 2 above)
Comment (smjn) [reply coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/MultiThreadedEventProcessorTest.java]: No, remnant of automated refactoring. Will fix

File: settings.gradle (hunk 5 above)
Comment (smjn) [reply settings.gradle:60]: AFAIK, all the subprojects/modules for the kafka project must be declared in settings.gradle
Don't think this has anything to do with `clients` module.

**Comment by smjn** [comment]
> thanks! left some comments.
> 
> also, i was wondering if CoordinatorRecordSerde and CoordinatorRecordSerde should now be prefixed with `Group`

We can - but they are already in `group` package, hence might be redundant.

File: coordinator-common/src/main/java/org/apache/kafka/coordinator/common/metrics/CoordinatorRuntimeMetrics.java [hunk 6]
```diff
@@ -14,9 +14,9 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.kafka.coordinator.group.metrics;
+package org.apache.kafka.coordinator.common.metrics;
 
-import org.apache.kafka.coordinator.group.runtime.CoordinatorRuntime.CoordinatorState;
+import org.apache.kafka.coordinator.common.runtime.CoordinatorRuntime.CoordinatorState;
 
 import java.util.function.Supplier;
```
Comment (AndrewJSchofield) [inline coordinator-common/src/main/java/org/apache/kafka/coordinator/common/metrics/CoordinatorRuntimeMetrics.java]: The comment on line 24 would be better if it just said "the coordinator runtimes".

**Review by AndrewJSchofield — APPROVED** [review]
Thanks for the PR. That looks really quite neat and shows that the idea of coordinator framework for multiple coordinators to use was done well.

File: checkstyle/import-control-coordinator-common.xml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+<import-control pkg="org.apache.kafka">
+
+    <!-- THINK HARD ABOUT THE LAYERING OF THE PROJECT BEFORE CHANGING THIS FILE -->
+
+    <!-- common library dependencies -->
+    <allow pkg="java" />
+    <allow pkg="org.slf4j" />
+    <allow pkg="org.junit" />
+    <allow pkg="org.mockito" />
+    <!-- no one depends on the server -->
```
Comment (apoorvmittal10) [inline checkstyle/import-control-coordinator-common.xml:30]: Should it be `no one depends on the coordinator-common`?

File: checkstyle/import-control-coordinator-common.xml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    <disallow pkg="kafka" />
+
+    <!-- anyone can use public classes -->
+    <!-- anyone can use public classes -->
```
Comment (apoorvmittal10) [inline checkstyle/import-control-coordinator-common.xml]: Are they 2 repeated comments?

**Review by apoorvmittal10 — COMMENTED** [review]
You also might want to consider correcting suppressions.xml to have classes under coordinator-runtime classification. Though it doesn't make any logical difference but just categorization.

File: checkstyle/import-control-coordinator-common.xml (hunk 7 above)
Comment (smjn) [reply checkstyle/import-control-coordinator-common.xml:30]: @apoorvmittal10 
No, that might be incorrect - other modules will depend on `coordinator-common`. Actually this comment exists in almost all checkstyle files.
Basically, we are disallowing accessing kafka package such that `import org.apache.kafka.ABCD` is not allowed and this is not module specific.

**Comment by smjn** [comment]
> You also might want to consider correcting suppressions.xml to have classes under coordinator-runtime classification. Though it doesn't make any logical difference but just categorization.

Could you please clarify? The PR does not introduce any change in suppressions.xml.

File: checkstyle/import-control-coordinator-common.xml (hunk 7 above)
Comment (apoorvmittal10) [reply checkstyle/import-control-coordinator-common.xml:30]: I think you are right 👍

File: coordinator-common/src/main/java/org/apache/kafka/coordinator/common/CoordinatorRecord.java [hunk 9]
```diff
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.kafka.coordinator.group;
+package org.apache.kafka.coordinator.common;
```
Comment (dajac) [inline coordinator-common/src/main/java/org/apache/kafka/coordinator/common/CoordinatorRecord.java]: I wonder if we should move `CoordinatorRecord`, `CoordinatorMetrics`, `CoordinatorMetricsShard` and `CoordinatorRuntimeMetrics` to the `runtime` package too. They are all generic classes which are part of the runtime framework in my opinion. What do you think?

File: coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java [hunk 10]
```diff
@@ -1775,7 +1775,7 @@ public void onHighWatermarkUpdated(
     /**
      * 16KB. Used for initial buffer size for write operations.
      */
-    static final int MIN_BUFFER_SIZE = 16384;
+    public static final int MIN_BUFFER_SIZE = 16384;
```
Comment (dajac) [inline coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java]: Is this change necessary?

File: coordinator-common/src/test/java/org/apache/kafka/coordinator/common/MockCoordinatorTimer.java [hunk 11]
```diff
@@ -43,7 +43,7 @@ public static class ScheduledTimeout<T, U> {
         public final long deadlineMs;
         public final TimeoutOperation<T, U> operation;
 
-        ScheduledTimeout(
+        public ScheduledTimeout(
```
Comment (dajac) [inline coordinator-common/src/test/java/org/apache/kafka/coordinator/common/MockCoordinatorTimer.java]: Are those changes necessary?

Comment (smjn) [reply coordinator-common/src/test/java/org/apache/kafka/coordinator/common/MockCoordinatorTimer.java]: Yes, it is used in group-coordinator

File: coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java (hunk 10 above)
Comment (smjn) [reply coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java]: No, it is only required within coordinator-common as of now

File: coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRecord.java [hunk 12]
```diff
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.kafka.coordinator.group;
+package org.apache.kafka.coordinator.common.runtime;
```
Comment (dajac) [inline coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRecord.java:17]: Are you going to reuse CoordinatorRecord for the share coordinator? If not, it may be better to keep it in the group-coordinator module for now as it may be specific to the implementation of the coordinator.

Comment (smjn) [reply coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRecord.java:17]: I will use it. The class is very general. I have moved the test class for the same as well.

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Review by AndrewJSchofield — APPROVED** [review]
Latest changes lgtm

**Review by apoorvmittal10 — APPROVED** [review]
Thanks for the changes. Suppressions.xml mentions `CoordinatorRecordHelpers` but the file has been renamed to `GroupCoordinatorRecordHelpers`. As the suppressions seems not to be a strict class file name match hence build should work but maybe to avoid confusion we might want to consider using explicit name in suppressions. I leave it on you.

**Review by apoorvmittal10 — APPROVED** [review]
Thanks @smjn for the additional fix.

**Review by jeffkbkim — APPROVED** [review]
LGTM!

File: settings.gradle (hunk 5 above)
Comment (chia7712) [inline settings.gradle:60]: I guess the next coordinator-related module is called `share-coordinator`, and there is existent module called `group-coordinator`. Hence, have we consider the name: `common-coordinator`? Or you want to align the naming with `server-common`

**Review by chia7712 — COMMENTED** [review]
@smjn this is a great refactor. overall LGTM. one small question is left. PTAL

File: settings.gradle (hunk 5 above)
Comment (smjn) [reply settings.gradle:60]: Mainly 2 reasons for the current name.

1. The new parent coordinator module does not do anything useful on its own. `common-coordinator` has the connotation that this module serves a well defined purpose like group coordinator but that is not the case. It just runs the framework classes and the useful work is only done when a cooperating coordinator provides concrete implementations. You are also right in pointing out the server-common nomenclature. But in the same spirit common-server has a very different connotation and seems incorrect.
2. The original author of the group-coordinator suggested the name.

Comment (chia7712) [reply settings.gradle:60]: @smjn thanks for sharing. sounds good

**Review by chia7712 — APPROVED** [review]
LGTM. the failed tests pass on my local.
```
./gradlew cleanTest :streams:test --tests EosIntegrationTest.shouldNotViolateEosIfOneTaskFailsWithState --tests RestoreIntegrationTest.shouldInvokeUserDefinedGlobalStateRestoreListener :tools:test --tests GetOffsetShellTest.testTopicPartitionsArg --tests MetadataQuorumCommandTest.testDescribeQuorumReplicationSuccessful --tests MetadataQuorumCommandTest.testDescribeQuorumStatusSuccessful :storage:test --tests ReassignReplicaMoveTest.executeTieredStorageTest --tests TransactionsWithTieredStoreTest.testSendOffsetsToTransactionTimeout --tests TransactionsWithTieredStoreTest.testAbortTransactionTimeout --tests TransactionsWithTieredStoreTest.testCommitTransactionTimeout :connect:runtime:test --tests ConnectWorkerIntegrationTest.testRequestTimeouts :metadata:test --tests KRaftMigrationDriverTest.testNoDualWriteBeforeMigration --tests QuorumControllerTest.testBrokerHeartbeatDuringMigration :core:test --tests ZkMigrationFailoverTest.testDriverSkipsEventsFromOlderEpoch --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testProduceConsumeViaAssign --tests PlaintextProducerSendTest.testValidBeforeAndAfterTimestampsWithinThreshold
```
